File: src/main/java/core/procedures/strings/ReplaceFirst.java
package core.procedures.strings;

import core.exceptions.WrongTypeException;
import core.procedures.AFn;
import core.procedures.FnArgsBuilder;

import java.util.regex.Pattern;

public final class ReplaceFirst extends AFn {

  public ReplaceFirst() {
    super(new FnArgsBuilder().min(3).max(3).build());
  }

  @Override
  public String getName() {
    return "replace-first";
  }

  @Override
  public String apply3(Object arg1, Object arg2, Object arg3) {
    if (!(arg1 instanceof CharSequence)) {
      throw new WrongTypeException(getName(), "String", arg1);
    }
    if (arg2 instanceof CharSequence && arg3 instanceof CharSequence) {
      return arg1.toString().replaceFirst(arg2.toString(), arg3.toString());
    }
    // TODO arg2=string/arg3=function of match
    return ((Pattern)arg2).matcher((CharSequence)arg1).replaceFirst(((CharSequence)arg3).toString());
  }
}

File: src/main/java/core/procedures/strings/ReplaceFirst.kt
package core.procedures.strings

import core.exceptions.WrongTypeException
import core.procedures.AFn

import java.util.regex.Pattern

class ReplaceFirst : AFn<Any?, String>(name = "replace-first", isPure = true, minArgs = 3, maxArgs = 3) {

    override operator fun invoke(arg1: Any?, arg2: Any?, arg3: Any?): String {
        val chars = arg1 as? CharSequence ?: throw WrongTypeException(name, "String", arg1)
        if (arg2 is CharSequence && arg3 is CharSequence) {
            return chars.toString().replaceFirst(arg2.toString(), arg3.toString())
        }
        // TODO arg2=string/arg3=function of match
        return (arg2 as Pattern).matcher(chars).replaceFirst((arg3 as CharSequence).toString())
    }
}
